### ISSUE TITLE
type checking failed for List[Tuple[str] | Tuple[()]]

### ISSUE THREAD

**Issue by Freed-Wu** [microsoft/pyright#3836]
**To Reproduce**

```python
"""/dev/shm/test.py
❯ pyright test.py
No configuration file found.
No pyproject.toml file found.
stubPath /dev/shm/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
pyright 1.1.264
/dev/shm/test.py
  /dev/shm/test.py:26:7 - error: Argument of type "list[tuple[()]]" cannot be assigned to parameter "scaler_names" of type "List[Tuple[str] | Tuple[()]]" in function "test1"
    TypeVar "_T@list" is invariant
      Type "tuple[()]" cannot be assigned to type "Tuple[str] | Tuple[()]"
        Element size mismatch; expected 1 but received 0 (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
Completed in 0.491sec
"""
from typing import TYPE_CHECKING, List, Tuple


def test1(scaler_names: List[Tuple[str] | Tuple[()]]):
    print(scaler_names)


scaler_names = [()]
test1(scaler_names)  # error


def test2(scaler_names: Tuple[str] | Tuple[()]):
    print(scaler_names)


scaler_names = ()
test2(scaler_names)  # no error


def test3(scaler_names: List[Tuple[()]]):
    print(scaler_names)


scaler_names = [()]
test3(scaler_names)  # no error
```

**Expected behavior**

No error for pyright's result.

**Screenshots or Code**

![Screenshot from 2022-08-17 17-13-58](https://user-images.githubusercontent.com/32936898/185081981-bef4dc86-86bc-4a42-ab9c-3593c839858c.png)

**VS Code extension or command-line**

```shell
❯ has pyright
✔ pyright 1.1.264
```

**Comment by erictraut** [comment]
This is expected behavior, not a bug. The variable `scaler_names` has no declared type, so the normal inference rules for lists and tuples apply when evaluating the type of the expression `[()]`. These inference rules result in the type `List[Tuple[()]`. This type is not type compatible with the type `List[Tuple[str] | Tuple[()]]` because the type parameter for list is invariant.

You have a couple of options:
1. In `test1`, use a `Sequence` rather than a `list`.
2. Provide a declared type for variable `scalar_names`. Bidirectional type inference will then override the normal inference rules.
```python
def test1(scaler_names: List[Tuple[str] | Tuple[()]]):
    print(scaler_names)


scaler_names: List[Tuple[str] | Tuple[()]] = [()]
test1(scaler_names)
```

For more details about type inference, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#bidirectional-type-inference-expected-types).

**Comment by Freed-Wu** [comment]
You are right. Thanks!

**Comment by s-banach** [comment]
Can you help me understand why this is correct behavior?

```python
def f(x: "str | None"):
    pass

x = str(3)
f(x)  # Argument of type string is assigned to parameter of type str | None without complaint.
```

```python
def g(xs: "list[str | None]"):
    pass

xs = [str(3)]
g(xs)  #  Pyright error message: "Type str cannot be assigned to type str | None".
```

To me, it is obviously type-safe to pass a list of ints to a function that accepts a list of Optional ints.
